hasktorch/src/Torch/Static/NN.hs

```diff
@@ -0,0 +1,75 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Torch.Static.NN where
+
+import Control.Monad.State.Strict
+
+import GHC.TypeLits
+import GHC.Generics
+import Torch.Static
+import Torch.Static.Factories
+import Torch.Static.Native
+import qualified Torch.NN as A
+import qualified Torch.Autograd as A
+import qualified Torch.Tensor as A
+import qualified Torch.DType as D
+
+newtype Parameter dtype shape = Parameter A.IndependentTensor deriving (Show)
+
+toDependent :: Parameter dtype shape -> Tensor dtype shape
+toDependent (Parameter t) = UnsafeMkTensor $ A.toDependent t
+
+instance A.Parameterized (Parameter dtype shape) where
+  flattenParameters (Parameter x) = [x]
+  replaceOwnParameters _ = Parameter <$> A.nextParameter
+
+data LinearSpec (dtype :: D.DType) (inputFeatures :: Nat) (outputFeatures :: Nat) = LinearSpec
+  deriving (Show, Eq)
+
+data Linear (dtype :: D.DType) (inputFeatures :: Nat) (outputFeatures :: Nat) =
+  Linear { weight :: Parameter dtype '[inputFeatures, outputFeatures]
+         , bias :: Parameter dtype '[outputFeatures]
+         } deriving (Show, Generic)
+
+linear
+  :: forall dtype (inputFeatures :: Nat) (outputFeatures :: Nat) (shape :: [Nat]) (shape' :: [Nat])
+   . ( CheckBroadcast (CheckMatMul
+                         shape
+                         '[inputFeatures, outputFeatures]
+                         (ComputeMatMul
+                            (ReverseImpl shape '[]) '[outputFeatures, inputFeatures]))
+                      '[outputFeatures]
+                      (ComputeBroadcast
+                         (ReverseImpl
+                            (CheckMatMul
+                               shape
+                               '[inputFeatures, outputFeatures]
+                               (ComputeMatMul
+                                  (ReverseImpl shape '[]) '[outputFeatures, inputFeatures]))
+                            '[])
+                         '[outputFeatures])
+                    ~ shape')
```

Comment:
All of this should be possible to infer from the body. Why not use partial type signatures??